drgn/helpers/linux/module.py

```diff
@@ -0,0 +1,228 @@
+    mem = mod.mem.read_()
+    regions = []
+    for i in range(num_types.value_()):
+        if mem[i].size:
+            regions.append(_range_from_module_layout(mem[i]))
+    return regions
+
+
+def module_address_regions(mod: Object) -> List[Tuple[int, int]]:
+    """
+    Returns a list of address ranges for a module
+
+    Given a ``struct module *``, return every address range associated with the
+    module. Note that the number of address ranges and their interpretations
+    vary across kernel versions. Some kernel versions provide additional
+    information about some regions (e.g. text, data, R/O, init). This API
+    doesn't distinguish. However, this API does not provide the module's percpu
+    region: use ``module_percpu_region()`` for that.
+
+    :param mod: Object of type ``struct module *``
+    :returns: A list of address range tuples
+    """
+    try:
+        # Since Linux 6.4, ac3b432839234 ("module: replace module_layout with
+        # module_memory"), module address regions are broken into several types,
+        # each with their own base and size.
+        mod.prog_.constant("MOD_MEM_NUM_TYPES")
+    except LookupError:
+        pass
+    else:
+        return _ranges_from_module_memory(mod)
+
+    try:
+        # Prior to 6.4, there were two "struct module_layout" objects,
+        # core_layout and init_layout, which contained the module's memory
+        # layout and any memory which could be freed after init. The init_layout
+        # is usually NULL / size 0. The module_layout structure has more
+        # information to say where text ends, where rodata ends, etc. We ignore
+        # these.
+        core = _range_from_module_layout(mod.core_layout)
+        init = _range_from_module_layout(mod.init_layout)
+    except AttributeError:
+        # Prior to 4.5, 7523e4dc5057 ("module: use a structure to encapsulate
+        # layout."), the layout information was stored as variables directly in
+        # the struct module. They were prefixed with "core_" and "init_".
+        core = _range_from_module(mod, "core")
+        init = _range_from_module(mod, "init")
+
+    ret = [core]
+    if init:
+        ret.append(init)
+    return ret
+
+
+def _addrmod_tree(mod_tree: Object, addr: int) -> Object:
+    prog = mod_tree.prog_
+
+    # The module tree is "latched": there are two parallel trees. Which one is
+    # in use depends on the seqcount, which gets incremented for each
+    # modification. This is a really neat approach that allows reads in parallel
+    # with a writer. In our use case, it's probably not worth verifying the
+    # seqcount after the fact. What we do need is the index (0 or 1). This may
+    # be a seqcount_latch_t, or before 24bf401cebfd6 ("rbtree_latch: Use
+    # seqcount_latch_t"), a regular seqcount_t.
+    try:
+        idx = int(mod_tree.root.seq.seqcount.sequence & 1)
+    except AttributeError:
+        idx = int(mod_tree.root.seq.sequence & 1)
```

Comment:
I have a slight preference for:
```suggestion
        idx = mod_tree.root.seq.seqcount.sequence.value_() & 1
    except AttributeError:
        idx = mod_tree.root.seq.sequence.value_() & 1
```